### PR TITLE
New work calculation at block 600000 (soft fork)

### PR DESCRIPTION
**Pull request by MentalCollatz**
First off, this will fix the safe mode warnings once and for all.  But it does so much more than that.  Currently, an attacker can 51% attack the network with roughly 60% of SHA256D and nothing else.  After this change, an attacker with 90% of the SHA256D hashrate and 33% of each of the other 4 algorithms would have insufficient hashpower to mount a 51% attack.

The new formula was chosen as a function of the difficulties and based on these criteria:
1. It should be a symmetric funtion
2. It should be order 1 homogenous
3. It should be homogenous with respect to each variable

Or in plain English:
1. There should be nothing algorithm-specific (such as per-algorithm weights) nor should it depend on which algorithm actually solved the block.
2. If all difficulties double, the block work should double
3. If one difficulty doubles, the block work should change by some constant factor

There is only one function (save multiplication by a constant factor) satisfying all 3 conditions: the geometric mean.  As an added bonus, because of how the difficulty algorithm works the geometric mean can change by at most 3% from block to block (which addresses the safe-mode warning issue).

In order to 51% attack the network, the product of the attacker's hashrates must exceed the product of the network's hashrates.  In particular, the attacker must have some hashrate in all 5 algorithms.

### CODE REVIEW

**Comment by terrylai** [comment]
Hello, MentalCollatz,

I am Terry, an employee of Tofugear, which is a collaborative company with Digbyte. I am new to analyse the theory and implementation of digibyte core code. So the following saying might be my misunderstanding. Please feel free to point out that if I am wrong.

From Digibyte source code, 

https://github.com/digibyte/digibyte/blob/b43ee15714eadf5cae8e4a3a7ca65f5732c052f8/src/main.h#L862

the final work value has been taken the average.

The GetBlockWorkAdjusted function, which returns GetBlockWork()*GetAlgoWorkFactor()
and the GetAlgoWorkFactor() function gives a factor so that every final value from GetBlockWorkAdjusted will give a reasonable average work value.

So in conclusion it might not be a big help for your change, though it takes the real mean.

Besides, I want to ask that why in src/main.h line 888,
-            bnRes <<= 7;

you multiply the value by shifting left 7 bits? Is it to make the value somewhat near the old values before change?

BR,
Terry

**Comment by MentalCollatz** [comment]
Hi Terry,

The problem with GetAlgoWorkFactor() is that it it is based on heuristics that change over time.  For example the work factor for Skein is currently much lower than it should be and the work factor for SHA2d is much higher than it should be.  This results in 2 major problems:
1. The total "work" contributed by SHA2d is greater than the other 4 algorithms combined.  Consequently, SHA256d is weak against 51% attacks.
2. Safe mode can be triggered on a single-block fork.  The warning conditions are basically looking for a fork where a lot of work has been done, using the current tip as a reference point.  But the work-per-block can vary so wildly that we see a lot of false positives on the check.

Shifting the value left by 7 is to protect DigiByte from 51% attacks during the transition to the new work formula.  Without it, an attacker with just a couple ASICs would be able to start a fork a few blocks before the transition and later broadcast it and undo thousands of legitimate blocks.

MC

**Comment by terrylai** [comment]
Hello, MentalCollatz,

I am Terry again. Recently I am testing your new fork suggestion. 

I have tested scrypt algo and will continue to test the other algos. It is working well.

However, I have one question about your nth root function.

You are using approximation method to find the result.

I remembered that for approximation methods such as Newton's method are not guaranteed to be convergent, and it is mentioned in this following link.

http://en.wikipedia.org/wiki/Newton%27s_method

You have stated the 3 criteria, is it for the reason of convergence?

If not and once the result is not convergent, I studied that you have not had special handling in the function such as using old work value.

Please feel free to discuss this so as to keep safe and strict the mining job.

Have a nice weekend!

BR,
Terry

**Comment by MentalCollatz** [comment]
Hi Terry,

Regarding the convergence of Netwon's method, things change when you restrict the values to integers.  The wikipedia article only considers arbitrary precision arithmetic.

It remains important to have a good starting approximation.  During testing I found that the function did not always converge if I used a 1-bit starting approximation.  Sometimes it would oscillate between 2 values close to the answer.  However, using a 2-bit approximation converged in every test case.  The final version uses an 8-bit approximation.  I ran hours of stress tests and never hit a case in which the number of loop iterations ever exceeded 10.  The loop condition is just defensive programming.

MC

**Comment by ghost** [comment]
Hi MentalCollatz,

1. I want to understand WHY you believe "an attacker with 90% of the SHA256D hashrate and 33% of each of the other 4 algorithms would have insufficient hashpower to mount a 51% attack". I don't see why this claim is true, neither from the explanation in this pull request or the source code in it.
2. I saw no testcases that convincingly reproduce these condititions: a testcase that attempts this attack.
3. You said "I ran hours of stress tests and never hit a case in which the number of loop iterations ever exceeded 10." I didn't see a copy of these testcases in the code base. How can I run them? I also didn't see a copy of the testcase output for all these hours.

I also have a question, because I don't know and want to understand. If this type of security is so groundbreaking, why do practically zero other cryptocurrencies (and researchers, and mathematicians, and software engineers) use it?

Chime in when you're ready https://www.reddit.com/r/Digibyte/comments/8v4u9r/why_does_in_digibyte_an_attacker_with_90_of_the/

Thanks.
